doc/source/file_formats.rst

```diff
@@ -1,23 +1,375 @@
+
+For its description of the force fields, topologies and mappings,
+Vermouth uses a file format based on the one used by Gromacs,
+consisting of named sections and subsections, each indicated with a
+tag between square brackets. Tags are divided into top-level and
+sub-level section names.
+
+Force fields and topologies
+---------------------------
+
+The top-level tags for the force field and the topologies are macros,
+variables, citations, moleculetypes, links, and modifications.
+
+Macros
+~~~~~~
+
+The macros section has no further subsections and lists substitution
+patterns to be applied throughout the file being read. Macro values
+are substituted using the name with a preceding $. This is similar to
+the use of variables in shell scripting and makes it easy to write
+generalizations and to use and change default values. An example of
+this section is
+
+[ macros ]
+prot_default_bb_type P2
+stiff_fc 1000000
+
+Variables
+~~~~~~~~~
+
+The variables section has no further subsections and lists control
+parameters for the force field. An example of this is
+
+[ variables ]
+elastic_network_bond_type 1
+res_min_dist 3
+
+Citations
+~~~~~~~~~
+
+The citations section lists the citations to be used for the force
+field. Citations are named and refer to an entry in the bibtex file
+`citations.bib` in the force field directory. An example of this is
+
+[ citations ]
+Martini3
+
+Citations can also be specified for moleculetypes, links, and
+modifications, in a citation subsection.
+
+Moleculetypes
+~~~~~~~~~~~~~
+
+The moleculetype describes a :ref:`block <data:block>`, i.e., a
+topological building block, comprising of particles (atoms) with their
+properties and interactions. This can be a separate molecule or a part
+of a larger molecule, typically a monomeric unit in a polymer. The
+moleculetype has a name and the number of bonds to use for exclusions
+as its first content line. This is followed by one or more
+subsections. The subsection atoms is mandatory, optional subsections
+are edges and subsections corresponding to the different types of
+atomic interactions, including bonds, constraints, angles, and
+dihedrals.
+
+Subsection metadata
+^^^^^^^^^^^^^^^^^^^
+
+Subsections can be given metadata, using a #meta directive. These
+directives always apply to the whole subsection. This makes it
+possible to, e.g., group interactions or annotate bonds according to
+their type. It is also possible to specify a context for the
+interactions in a subsection, using 'ifdef' and/or 'ifndef' tags. This
+will make the inclusion of the interactions conditional by adding
+#ifdef and #ifndef directive statements to the output topology
+files. The metadata is given as a JSON style mapping of key/value
+pairs.
+
+#meta {"group": "Side chain bonds", "ifdef": "FLEXIBLE"}
+
+Metadata can also be added to a single line by adding an attribute
+statement as the last element.
+
+Atoms
+^^^^^
+
+Each line in the atoms section describes one particle, corresponding
```

Comment:
That's possible with the current format, no? The only restriction is that names have to be unique. It would be doable to say the combination of atom name and residue id+name needs to be unique, but we can't deal with all equal atom names.

Reply:
It's been a while since I've made datafiles but I'm sure there's some gotchas when you try to add a block from an itp file. For sure there is the requirement that atomnames must be unique (which probably produces an obscure error message). I'm not sure there's more. 
It's not really an issue if you add small molecules/residues (the main usecase). It blows up in your face for polymers of course.
Is it desirable to have the parser generate unique atomnames by adding numbers? I kind of doubt it, since that complicates writing mappings (and links and modifications) since you can no longer lookup atom names in the itp file. An intelligible warning/error would be my preference I think.